File: src/packages/control-ui/src/api/desktop_client_manager.ts
export enum SystemProxyMode {
  Direct = 'direct',
  Pac = 'pac',
  Global = 'global',
  Manual = 'manual',
}

export interface DesktopStatus {
  mode: SystemProxyMode;
  activeProfile: string;
  httpPort: number;
  socks5Port: number;
  isConnected: boolean;
  pacUrl?: string;
}

export class DesktopClientManager {
  private mode: SystemProxyMode = SystemProxyMode.Direct;
  private activeProfile: string = 'default';
  public httpPort: number;
  public socks5Port: number;
  private isConnected: boolean = false;
  private pacUrl?: string;

  constructor(httpPort: number, socks5Port: number) {
    this.httpPort = httpPort;
    this.socks5Port = socks5Port;
  }

  setProxyMode(mode: SystemProxyMode, pacUrl?: string): void {
    if (mode === SystemProxyMode.Pac && !pacUrl && !this.pacUrl) {
      throw new Error('PAC mode requires a valid PAC URL');
    }

    this.mode = mode;
    if (pacUrl) {
      this.pacUrl = pacUrl;
    }
  }

  switchProfile(profileId: string): void {
    this.activeProfile = profileId;
  }

  setConnected(connected: boolean): void {
    this.isConnected = connected;
  }

  getStatus(): DesktopStatus {
    return {
      mode: this.mode,
      activeProfile: this.activeProfile,
      httpPort: this.httpPort,
      socks5Port: this.socks5Port,
      isConnected: this.isConnected,
      pacUrl: this.pacUrl,
    };
  }

  handleIpcCommand(command: string, payload: Record<string, any> = {}): any {
    switch (command) {
      case 'get_status':
        return this.getStatus();
      case 'set_mode': {
        const modeStr = payload.mode;
        if (!modeStr) throw new Error('Missing mode');
        let mode: SystemProxyMode;
        switch (modeStr.toLowerCase()) {
          case 'direct':
            mode = SystemProxyMode.Direct;
            break;
          case 'pac':
            mode = SystemProxyMode.Pac;
            break;
          case 'global':
            mode = SystemProxyMode.Global;
            break;
          case 'manual':
            mode = SystemProxyMode.Manual;
            break;
          default:
            throw new Error('Invalid proxy mode');
        }
        this.setProxyMode(mode, payload.pac_url || payload.pacUrl);
        return { success: true };
      }
      case 'switch_profile': {
        const profile = payload.profile;
        if (!profile) throw new Error('Missing profile');
        this.switchProfile(profile);
        return { success: true, profile };
      }
      default:
        throw new Error(`Unknown command: ${command}`);
    }
  }
}

File: src/packages/lumicore/src/platform/desktop_client_manager.rs
//! # Desktop Client Manager
//!
//! Cross-platform desktop proxy controller, system proxy mode supervisor,
//! and IPC event router for desktop client frontends.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum SystemProxyMode {
    Direct,
    Pac,
    Global,
    Manual,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DesktopStatus {
    pub mode: SystemProxyMode,
    pub active_profile: String,
    pub http_port: u16,
    pub socks5_port: u16,
    pub is_connected: bool,
    pub pac_url: Option<String>,
}

pub struct DesktopClientManager {
    mode: SystemProxyMode,
    active_profile: String,
    http_port: u16,
    socks5_port: u16,
    is_connected: bool,
    pac_url: Option<String>,
}

impl DesktopClientManager {
    pub fn new(http_port: u16, socks5_port: u16) -> Self {
        Self {
            mode: SystemProxyMode::Direct,
            active_profile: "default".to_string(),
            http_port,
            socks5_port,
            is_connected: false,
            pac_url: None,
        }
    }

    pub fn set_proxy_mode(&mut self, mode: SystemProxyMode, pac_url: Option<String>) -> Result<(), String> {
        if mode == SystemProxyMode::Pac && pac_url.is_none() && self.pac_url.is_none() {
            return Err("PAC mode requires a valid PAC URL".to_string());
        }

        self.mode = mode;
        if pac_url.is_some() {
            self.pac_url = pac_url;
        }
        Ok(())
    }

    pub fn switch_profile(&mut self, profile_id: &str) {
        self.active_profile = profile_id.to_string();
    }

    pub fn set_connected(&mut self, connected: bool) {
        self.is_connected = connected;
    }

    pub fn get_status(&self) -> DesktopStatus {
        DesktopStatus {
            mode: self.mode,
            active_profile: self.active_profile.clone(),
            http_port: self.http_port,
            socks5_port: self.socks5_port,
            is_connected: self.is_connected,
            pac_url: self.pac_url.clone(),
        }
    }

    pub fn handle_ipc_command(&mut self, command: &str, payload: serde_json::Value) -> Result<serde_json::Value, String> {
        match command {
            "get_status" => {
                let status = self.get_status();
                serde_json::to_value(status).map_err(|e| e.to_string())
            }
            "set_mode" => {
                let mode_str = payload.get("mode").and_then(|v| v.as_str()).ok_or("Missing mode")?;
                let pac_url = payload.get("pac_url").and_then(|v| v.as_str()).map(|s| s.to_string());
                let mode = match mode_str {
                    "direct" => SystemProxyMode::Direct,
                    "pac" => SystemProxyMode::Pac,
                    "global" => SystemProxyMode::Global,
                    "manual" => SystemProxyMode::Manual,
                    _ => return Err("Invalid proxy mode".to_string()),
                };
                self.set_proxy_mode(mode, pac_url)?;
                Ok(serde_json::json!({ "success": true }))
            }
            "switch_profile" => {
                let profile = payload.get("profile").and_then(|v| v.as_str()).ok_or("Missing profile")?;
                self.switch_profile(profile);
                Ok(serde_json::json!({ "success": true, "profile": profile }))
            }
            _ => Err(format!("Unknown command: {}", command)),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_desktop_client_manager() {
        let mut mgr = DesktopClientManager::new(7890, 7891);
        let status = mgr.get_status();
        assert_eq!(status.mode, SystemProxyMode::Direct);
        assert!(!status.is_connected);

        let res = mgr.handle_ipc_command("set_mode", serde_json::json!({ "mode": "global" }));
        assert!(res.is_ok());
        assert_eq!(mgr.get_status().mode, SystemProxyMode::Global);

        let res2 = mgr.handle_ipc_command("switch_profile", serde_json::json!({ "profile": "us_east_node" }));
        assert!(res2.is_ok());
        assert_eq!(mgr.get_status().active_profile, "us_east_node");
    }
}
